source/server/connection_handler_impl.cc

```diff
@@ -39,60 +39,84 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
     ConnectionHandler::ActiveUdpListenerPtr udp_listener =
         config.udpListenerConfig()->listenerFactory().createActiveUdpListener(*worker_index_, *this,
                                                                               dispatcher_, config);
-    details.typed_listener_ = *udp_listener;
-    details.listener_ = std::move(udp_listener);
+    details->typed_listener_ = *udp_listener;
+    details->listener_ = std::move(udp_listener);
   }
+
   if (disable_listeners_) {
-    details.listener_->pauseListening();
+    details->listener_->pauseListening();
   }
-  if (auto* listener = details.listener_->listener(); listener != nullptr) {
+  if (auto* listener = details->listener_->listener(); listener != nullptr) {
     listener->setRejectFraction(listener_reject_fraction_);
   }
-  listeners_.emplace_back(config.listenSocketFactory().localAddress(), std::move(details));
+
+  details->listener_tag_ = config.listenerTag();
+  details->address_ = config.listenSocketFactory().localAddress();
+
+  ASSERT(!listener_map_by_tag_.contains(config.listenerTag()));
+
+  listener_map_by_tag_.emplace(config.listenerTag(), details);
+  // This map only store the new listener.
+  if (absl::holds_alternative<std::reference_wrapper<ActiveTcpListener>>(
+          details->typed_listener_)) {
+    tcp_listener_map_by_address_.insert_or_assign(
+        config.listenSocketFactory().localAddress()->asStringView(), details);
+  } else if (absl::holds_alternative<std::reference_wrapper<ActiveInternalListener>>(
+                 details->typed_listener_)) {
+    internal_listener_map_by_address_.insert_or_assign(
+        config.listenSocketFactory().localAddress()->asStringView(), details);
+  }
 }
 
 void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
-  for (auto listener = listeners_.begin(); listener != listeners_.end();) {
-    if (listener->second.listener_->listenerTag() == listener_tag) {
-      listener = listeners_.erase(listener);
-    } else {
-      ++listener;
+  if (auto listener_iter = listener_map_by_tag_.find(listener_tag);
+      listener_iter != listener_map_by_tag_.end()) {
+    // listener_map_by_address_ may already update to the new listener. Compare it with the one
+    // which find from listener_map_by_tag_, only delete it when it is same listener.
+    auto address_view = listener_iter->second->address_->asStringView();
+    if (tcp_listener_map_by_address_.contains(address_view) &&
+        tcp_listener_map_by_address_[address_view]->listener_tag_ ==
+            listener_iter->second->listener_tag_) {
+      tcp_listener_map_by_address_.erase(address_view);
+    } else if (internal_listener_map_by_address_.contains(address_view) &&
+               internal_listener_map_by_address_[address_view]->listener_tag_ ==
+                   listener_iter->second->listener_tag_) {
+      internal_listener_map_by_address_.erase(address_view);
     }
+    listener_map_by_tag_.erase(listener_tag);
```

Comment:
It's more efficient to erase with `listener_iter` instead of by tag, so that we don't have to find the element again.

Reply:
got it